go-controller/pkg/ovn/controller/apbroute/external_controller_policy.go

```diff
@@ -430,525 +293,153 @@ func (m *externalPolicyManager) calculateAnnotatedPodGatewayIPsForNamespace(targ
-	if err != nil {
-		return nil, nil, nil, err
+	if dynamicGWInfo.Len() > 0 {
+		klog.V(5).Infof("Found dynamic hops for policy %s: %+v", policy.Name, dynamicGWInfo)
 	}
 
-	return mismatchingNamespaces, mismatchingStaticHops, mismatchingDynamicHops, nil
-}
+	targetNs, err := m.listNamespacesBySelector(&policy.Spec.From.NamespaceSelector)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list target namespaces: %w", err)
+	}
 
-// calculateNamespaceSelectorDifferences determines the difference between the first and the second selector. The outcome is a set that contains
-// those namespace names that are in the first selector but not found in the second selector.
-func (m *externalPolicyManager) calculateNamespaceSelectorDifferences(nsSelector1, nsSelector2 *metav1.LabelSelector) (sets.Set[string], error) {
-	unmatchingNamespaces := sets.New[string]()
-	if !reflect.DeepEqual(nsSelector1, nsSelector2) {
-		nsList1, err := m.listNamespacesBySelector(nsSelector1)
-		if err != nil {
-			return nil, err
-		}
-		nsList2, err := m.listNamespacesBySelector(nsSelector2)
+	targetNsNames := sets.Set[string]{}
+	targetNamespaces := map[string]map[ktypes.NamespacedName]*v1.Pod{}
+	for _, ns := range targetNs {
+		targetNsNames.Insert(ns.Name)
+		targetPods, err := m.podLister.Pods(ns.Name).List(labels.Everything())
 		if err != nil {
-			return nil, err
+			return nil, fmt.Errorf("failed to get all ns %s pods: %v", ns.Name, err)
 		}
-		for _, ns1 := range nsList1 {
-			var found bool
-			for _, ns2 := range nsList2 {
-				if ns1.Name == ns2.Name {
-					found = true
-					break
-				}
-			}
-			if !found {
-				unmatchingNamespaces.Insert(ns1.Name)
+		podsMap := map[ktypes.NamespacedName]*v1.Pod{}
+		for _, pod := range targetPods {
+			// Ignore completed pods, host networked pods, pods not scheduled
+			if util.PodWantsHostNetwork(pod) || util.PodCompleted(pod) || !util.PodScheduled(pod) {
+				continue
 			}
+			podsMap[getPodNamespacedName(pod)] = pod
 		}
+		targetNamespaces[ns.Name] = podsMap
 	}
-	return unmatchingNamespaces, nil
-}
-
-// calculateStaticHopDifferences determines the difference between the first slice and the second staticHops slice. The outcome is a slice
-// of static hops that are in the staticHop1 slice but not in the staticHop2 slice.
-func (m *externalPolicyManager) calculateStaticHopDifferences(staticHops1, staticHops2 []*adminpolicybasedrouteapi.StaticHop) []*adminpolicybasedrouteapi.StaticHop {
-	diffStatic := make([]*adminpolicybasedrouteapi.StaticHop, 0)
-	for _, staticHop1 := range staticHops1 {
-		var found bool
-		for _, staticHop2 := range staticHops2 {
-			if reflect.DeepEqual(staticHop1, staticHop2) {
-				found = true
-				break
-			}
-		}
-		if !found {
-			diffStatic = append(diffStatic, staticHop1)
+	for policyName, refObjs := range m.policyReferencedObjects {
+		if policyName == policy.Name {
+			continue
 		}
-	}
-	return diffStatic
-}
+		if refObjs.targetNamespaces.Intersection(targetNsNames).Len() > 0 {
```

Comment:
it has a chance to succeed if another policy targeting this namespace is deleted or updated, and I think we even have a test that checks that case. May be useful e.g. if another policy was deleted, but that delete event wasn't handled yet, wdyt?

Reply:
If the current policy that is targeting the namespace changes, then the reconciliation in the `syncNamespace()` should queue both policies. But now I realize that if the processing order in the queue is to first tackle the failed policy, using the logic I just suggested will lead to a namespace without any policy, since there will not be any retry and the second policy will be removed afterwards.

I guess we have to keep retrying for this kind of situations.

Reply:
it's even a bit worse :)
namespace doesn't change in this scenario, so there will be no namespace events, and if second policy targeting the same namespace can't be configured in 15 retries, it will be removed from the queue, so then even if another policy targeting that namespace is deleted (which means the second policy can be configured now) there will be no events to trigger second policy update. But we can just document this I guess, advising users e.g. to recreate second policy based on status message